### ISSUE TITLE
pdist to accept dtype for float accuracy

### ISSUE THREAD

**Issue by idantene** [scipy/scipy#11280]
**Is your feature request related to a problem? Please describe.**
With array of e.g. 60,000, an array of 13.41GB is allocated. If the wrapper functions in `distance_wrap.c` would also allow e.g. `float32` and `float16`, that can be reduced to 3.35GB, which may suffice for most systems and allow more work to be done on a single node.

**Describe the solution you'd like**
From a user perspective, include a `dtype` argument in `pdist` and `cdist` functions, and/or allow an `out` array with a type different than `double`.

**Describe alternatives you've considered**
Implementing a pdist-like solution with `numba` and JIT/AOT compilation.

**Comment by rgommers** [comment]
Hi @idantene, this is not an unreasonable request, however there's at least two issues:

1. We cannot introduce `dtype=None` and drop the conversion to `float64` easily, because it will reduce precision for current use of `float32`/etc usage. So there's a backwards compat impact.
2. The C code in `scipy/spatial/src/` would need to be updated to support the new dtypes. This is a lot of work, everything would need to be templated and audited for numerical issues with reduced precision dtypes.

If you or someone else would like to work on this we can discuss; at the moment I'm leaning towards "let's not do this".

**Comment by idantene** [comment]
Hey @rgommers, thanks for the fast reply!
1. This could probably be resolved with a simple deprecated warning (i.e. introducing a default `dtype=float64` and alerting it would soon change). How does that sound?
2. The amount of templating seems to depend on which metrics should be included in this conversion. For example, all the metrics that operator on boolean vectors do not need to be changed.  This leaves us with `bray_curtis, canberra, chebyshev, city_block, euclidean, jaccard, jensenshannon, sqeuclidean`. 

I wouldn't mind looking into it myself, but I'm definitely missing some background on the numpy C-API and build process to begin with.

**Comment by rgommers** [comment]
>     1. This could probably be resolved with a simple deprecated warning (i.e. introducing a default `dtype=float64` and alerting it would soon change). How does that sound?

Yes, `dtype=float64` would work. We could either add a deprecation/future-warning, or just leave it at `float64`. The latter may avoid trouble, and users can change to `dtype=None` then to avoid the casting.

>     2. The amount of templating seems to depend on which metrics should be included in this update. For example, all the metrics that operator on boolean vectors do not need to be changed.  This leaves us with `bray_curtis, canberra, chebyshev, city_block, euclidean, jaccard, jensenshannon, sqeuclidean`. Am I missing something?

And `hamming`, `minkowski`, `wminkowski` it looks like. But perhaps it's indeed not too bad.

>     3. Could you elaborate on the numerical issues you're concerned with? (i.e. consequences of using `uint` and similar; what other datatypes could introduce numerical issues?)

Simple loss of precision. Looking at the implementations, it's mostly multiplies, running sums, dot products etc. - those tend to be more accurate when intermediate values are 64-bit.

> I wouldn't mind looking into it myself, but I'm definitely missing some background on the numpy C-API and build process to begin with.

Finishing off gh-5763 (translation to Cython) is probably the easiest way to go. There's also gh-3163 (change distance functions into gufuncs) but that's less complete.

Anyone else (@pv, @perimosocordiae?) have a preference on the way to go here?

Note that gh-9961 is related to this issue.

**Comment by perimosocordiae** [comment]
I would strongly prefer to wrap up the Cythonization cleanup PR before adding additional complexity here.

That said, I'm not opposed to providing reduced-precision distance functions so long as the maintenance burden isn't linear with the number of dtypes we support.

**Comment by idantene** [comment]
> I would strongly prefer to wrap up the Cythonization cleanup PR before adding additional complexity here.

Are you referring to [gh-5763](https://github.com/scipy/scipy/pull/5763)? Seems like there was no activity on it since 2017.

> That said, I'm not opposed to providing reduced-precision distance functions so long as the maintenance burden isn't linear with the number of dtypes we support.

Agreed. I'd imagine once these are templated for the type, then it won't be linear with number of dtypes.

**Comment by perimosocordiae** [comment]
Yeah, that PR has been dormant for a while, but the underlying idea is still sound. There's been some cleanup to the distance_wrap code since (by @apbard) so the rebase may look hairy.

**Comment by rgommers** [comment]
This is perhaps easier now that gh-13907 (reimplement cdist/pdist machinery in templated C++) is merged.

**Comment by lucascolley** [comment]
gh-19707 just requested this again, albeit via `scipy.cluster.hierarchy.linkage` and `seaborn.clustermap`.
